src/effects/effectprocessor.h

```diff
@@ -6,100 +7,208 @@
+    // callback executes process() with EffectEnableState::Disabling
+    virtual void deleteStatesForInputChannel(const ChannelHandle& inputChannel) = 0;
+
+    // Take a buffer of audio samples as pInput, process the buffer according to
+    // Effect-specific logic, and output it to the buffer pOutput. Both pInput
+    // and pOutput are represented as stereo interleaved samples for now, but
+    // effects should not be written assuming this will remain true. The properties
+    // of the buffer necessary for determining how to process it (frames per
+    // buffer, number of channels, and sample rate) are available on the
+    // mixxx::AudioParameters argument. The provided channel handles allow
+    // EffectProcessorImpl::process to fetch the appropriate EffectState and
+    // pass it on to EffectProcessorImpl::processChannel, allowing one
+    // EffectProcessor instance to process multiple signals simultaneously.
+    virtual void process(const ChannelHandle& inputHandle,
+                         const ChannelHandle& outputHandle,
                          const CSAMPLE* pInput, CSAMPLE* pOutput,
-                         const unsigned int numSamples,
-                         const unsigned int sampleRate,
-                         const enum EnableState enableState,
+                         const mixxx::AudioParameters& bufferParameters,
+                         const EffectEnableState enableState,
                          const GroupFeatureState& groupFeatures) = 0;
 };
 
-// Helper class for automatically fetching channel state parameters upon receipt
-// of a channel-specific process call.
-template <typename T>
-class PerChannelEffectProcessor : public EffectProcessor {
-    struct ChannelStateHolder {
-        ChannelStateHolder() : state(NULL) { }
-        T* state;
-    };
+// EffectProcessorImpl manages a separate EffectState for every routing of
+// input channel to output channel. This allows for processing effects in
+// parallel for PFL and post-fader for the master output.
+// EffectSpecificState must be a subclass of EffectState.
+template <typename EffectSpecificState>
+class EffectProcessorImpl : public EffectProcessor {
   public:
-    PerChannelEffectProcessor() {
-    }
-    virtual ~PerChannelEffectProcessor() {
-        for (typename ChannelHandleMap<ChannelStateHolder>::iterator it =
-                     m_channelState.begin();
-             it != m_channelState.end(); ++it) {
-            T* pState = it->state;
-            delete pState;
-        }
-        m_channelState.clear();
+    EffectProcessorImpl() {
     }
-
-    virtual void initialize(
-            const QSet<ChannelHandleAndGroup>& registeredChannels) {
-        foreach (const ChannelHandleAndGroup& channel, registeredChannels) {
-            getOrCreateChannelState(channel.handle());
+    // Subclasses should not implement their own destructor. All state should
```

Comment:
This is not true, even if you do not implement a destructor the effect classes will have an implicit destructor. 
So we should be save here and make the destructor virtual.

Reply:
Ok, it is already virtual by via the base class.  The override keyword is correct here.